lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUTransformCharFilter.java

```diff
@@ -0,0 +1,153 @@
+        "яяяяя", "âââââ"); // final NFC transform applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 0,
+        "яяяяя", "a\u0302a\u0302a\u0302a\u0302a\u0302"); // final NFC transform never applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 2,
+        "яяяяя", "ââa\u0302a\u0302a\u0302");
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 4,
+        "яяяяяяяяяя", "ââa\u0302a\u0302a\u0302a\u0302a\u0302a\u0302a\u0302a\u0302");
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 8,
+        "яяяяяяяяяяяяяяяяяяяя", "ââââââa\u0302ââââa\u0302ââââa\u0302âââ");
+  }
+
+  public void testCustomFunctionality() throws Exception {
+    String rules = "a > b; b > c;"; // convert a's to b's and b's to c's
+    checkToken(Transliterator.createFromRules("test", rules, Transliterator.FORWARD), "abacadaba", "bcbcbdbcb");
+  }
+  
+  public void testCustomFunctionality2() throws Exception {
+    String rules = "c { a > b; a > d;"; // convert a's to b's and b's to c's
+    checkToken(Transliterator.createFromRules("test", rules, Transliterator.FORWARD), "caa", "cbd");
+  }
+  
+  public void testOptimizer() throws Exception {
+    String rules = "a > b; b > c;"; // convert a's to b's and b's to c's
+    Transliterator custom = Transliterator.createFromRules("test", rules, Transliterator.FORWARD);
+    assertTrue(custom.getFilter() == null);
+    new ICUTransformCharFilter(new StringReader(""), custom);
+    assertTrue(custom.getFilter().equals(new UnicodeSet("[ab]")));
+  }
+  
+  public void testOptimizer2() throws Exception {
+    checkToken(Transliterator.getInstance("Traditional-Simplified; CaseFold"), 
+        "ABCDE", "abcde");
+  }
+  
+  public void testOptimizerSurrogate() throws Exception {
+    String rules = "\\U00020087 > x;"; // convert CJK UNIFIED IDEOGRAPH-20087 to an x
+    Transliterator custom = Transliterator.createFromRules("test", rules, Transliterator.FORWARD);
+    assertTrue(custom.getFilter() == null);
+    new ICUTransformCharFilter(new StringReader(""), custom);
+    assertTrue(custom.getFilter().equals(new UnicodeSet("[\\U00020087]")));
+  }
+
+  private void checkToken(Transliterator transform, String input, String expected) throws IOException {
+    checkToken(transform, ICUTransformCharFilter.DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY, input, expected);
+  }
+
+  private void checkToken(Transliterator transform, int maxRollbackBufferCapacity, String input, String expected) throws IOException {
+    final KeywordTokenizer input1 = new KeywordTokenizer();
+    input1.setReader(new ICUTransformCharFilter(new StringReader(input), transform, maxRollbackBufferCapacity));
+    assertTokenStreamContents(input1, new String[] { expected });
+  }
+  
+  /** blast some random strings through the analyzer */
+  public void testRandomStrings() throws Exception {
+    final Transliterator transform = Transliterator.getInstance("Any-Latin");
```

Comment:
What does this transformer do? Does it change the lengths of characters? Do we adequately test the many edge cases in the charfilter? Have you for example looked at code coverage from these tests?

Reply:
This maps any-character input to Latin output (e.g., "φῄς" => "phḗis" -- length 3 => length 5). So it often increases character length wrt input, but only occasionally (e.g., "ऄॻ" => "ॻ" -- length 2 => length 1) decreases character length wrt input (in this case -- and probably most/all others -- simply by deleting a character, it seems?).
What if we added a test that took a bidirectional CJK filter like "Katakana-Latin", mapped from a Kana dictionary to Latin, then tried to round-trip that transformation (back to Katakana). Both directions would be active transformations, aggressively manipulating the offset map (increasing and decreasing character count, respectively)?

Reply:
Got serviceable results by just running another `testRandomStrings()` with "Latin-Katakana". This resulted in plenty of positive diffs (decreased character count) to complement the negative diffs (decreased character counts) that were generated by the "Any-Latin" transform, so I think that'll suffice? (just pushed)